zhmccli/_cmd_console.py

```diff
@@ -363,17 +382,27 @@ def cmd_console_upgrade(cmd_ctx, options):
             "the Web Services API".format(v=hmc_version),
             cmd_ctx.error_format)
 
-    click.echo("Upgrading the HMC to bundle level {bl} and waiting for "
-               "completion (timeout: {t} s)".
-               format(bl=bundle_level, t=timeout))
+    fw_str = "FTP server {fs!r}".format(fs=ftp_host) if ftp_host else \
+        "the IBM support site"
+    click.echo("Upgrading the HMC to bundle level {bl} with firmware retrieval "
+               "from {fw} and waiting for completion (timeout: {t} s)".
+               format(bl=bundle_level, fw=fw_str, t=timeout))
+
+    kwargs = dict(
+        bundle_level=bundle_level,
+        accept_firmware=accept_firmware,
+        backup_location_type=backup_location_type,
+        wait_for_completion=True,
+        operation_timeout=timeout)
+    if ftp_host:
+        kwargs['ftp_host'] = ftp_host
+        kwargs['ftp_protocol'] = options['ftp_protocol']
+        kwargs['ftp_user'] = options['ftp_user']
+        kwargs['ftp_password'] = options['ftp_password']
```

Comment:
another assumption: this code gives a meaningful error when the user provided a ftp host, but user or password was forgotten?

Reply:
"click" ensures that the `options` dictionary has these items. If not specified on the command line, they will have the default value defined in the click options definition. The default default (if not specified there) is `None`.